ci/ci/ci.py

```diff
@@ -553,6 +566,85 @@ async def unfreeze_deploys(request, userdata):  # pylint: disable=unused-argumen
     return web.HTTPFound(deploy_config.external_url('ci', '/'))
 
 
+@routes.get('/namespaces')
+@auth.web_authenticated_developers_only()
+async def get_active_namespaces(request, userdata):
+    db: Database = request.app['db']
+    namespaces = [
+        r
+        async for r in db.execute_and_fetchall(
+            '''
+SELECT active_namespaces.*, JSON_ARRAYAGG(service) as services
+FROM active_namespaces
+LEFT JOIN deployed_services
+ON active_namespaces.namespace = deployed_services.namespace
+GROUP BY active_namespaces.namespace'''
+        )
+    ]
+    for ns in namespaces:
+        ns['services'] = [s for s in json.loads(ns['services']) if s is not None]
+    context = {
+        'namespaces': namespaces,
+    }
+    return await render_template('ci', request, userdata, 'namespaces.html', context)
+
+
+@routes.post('/namespaces/{namespace}/services/add')
+@check_csrf_token
+@auth.web_authenticated_developers_only()
+async def add_namespaced_service(request, userdata):  # pylint: disable=unused-argument
+    db: Database = request.app['db']
+    post = await request.post()
+    service = post['service']
+    namespace = request.match_info['namespace']
+
+    try:
+        await db.execute_insertone(
+            'INSERT INTO deployed_services VALUES (%s, %s)',
+            (namespace, service),
+        )
+    except pymysql.err.IntegrityError as err:
+        # 1062 ER_DUP_ENTRY https://dev.mysql.com/doc/refman/5.7/en/server-error-reference.html#error_er_dup_entry
+        if err.args[0] != 1062:
+            raise
+
+    return web.HTTPFound(deploy_config.external_url('ci', '/namespaces'))
+
+
+@routes.post('/namespaces/add')
+@check_csrf_token
+@auth.web_authenticated_developers_only()
+async def add_namespace(request, userdata):  # pylint: disable=unused-argument
+    db: Database = request.app['db']
+    post = await request.post()
+    namespace = post['namespace']
+
+    try:
+        await db.execute_insertone(
+            'INSERT INTO active_namespaces (`namespace`) VALUES (%s)',
+            (namespace,),
+        )
+    except pymysql.err.IntegrityError as err:
+        # 1062 ER_DUP_ENTRY https://dev.mysql.com/doc/refman/5.7/en/server-error-reference.html#error_er_dup_entry
+        if err.args[0] != 1062:
+            raise
+
+    return web.HTTPFound(deploy_config.external_url('ci', '/namespaces'))
+
+
+async def cleanup_expired_namespaces(db: Database):
+    expired_namespaces = [
```

Comment:
do we need an assert here that the DEFAULT NAMESPACE == "default"? I don't think this function is called in test deployments based on the if statement where we ensure the future for this function in the loop.

Reply:
I'll add it for peace of mind